examples/fabric/dcgan/train_torch.py

```diff
@@ -0,0 +1,271 @@
+nz = 100
+# Size of feature maps in generator
+ngf = 64
+# Size of feature maps in discriminator
+ndf = 64
+# Number of training epochs
+num_epochs = 5
+# Learning rate for optimizers
+lr = 0.0002
+# Beta1 hyperparameter for Adam optimizers
+beta1 = 0.5
+# Number of GPUs to use
+num_gpus = 1
+
+
+def main():
+    # Set random seed for reproducibility
+    seed = 999
+    print("Random Seed: ", seed)
+    random.seed(seed)
+    torch.manual_seed(seed)
+
+    dataset = CelebA(
+        root=dataroot,
+        split="all",
+        download=True,
+        transform=transforms.Compose(
+            [
+                transforms.Resize(image_size),
+                transforms.CenterCrop(image_size),
+                transforms.ToTensor(),
+                transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5)),
+            ]
+        ),
+    )
+
+    # Create the dataloader
+    dataloader = torch.utils.data.DataLoader(dataset, batch_size=batch_size, shuffle=True, num_workers=workers)
+
+    # Decide which device we want to run on
+    device = torch.device("cuda:0" if (torch.cuda.is_available() and num_gpus > 0) else "cpu")
+
+    output_dir = Path("outputs-torch", time.strftime("%Y%m%d-%H%M%S"))
+    output_dir.mkdir(parents=True, exist_ok=True)
+
+    # Plot some training images
+    real_batch = next(iter(dataloader))
+    torchvision.utils.save_image(
+        real_batch[0][:64],
+        output_dir / "sample-data.png",
+        padding=2,
+        normalize=True,
+    )
+
+    # Create the generator
+    generator = Generator().to(device)
+
+    # Handle multi-gpu if desired
+    if (device.type == "cuda") and (num_gpus > 1):
+        generator = nn.DataParallel(generator, list(range(num_gpus)))
+
+    # Apply the weights_init function to randomly initialize all weights
+    generator.apply(weights_init)
+
+    # Create the Discriminator
+    discriminator = Discriminator().to(device)
+
+    # Handle multi-gpu if desired
+    if (device.type == "cuda") and (num_gpus > 1):
+        discriminator = nn.DataParallel(discriminator, list(range(num_gpus)))
+
+    # Apply the weights_init function to randomly initialize all weights
+    discriminator.apply(weights_init)
+
+    # Initialize BCELoss function
+    criterion = nn.BCELoss()
+
+    # Create batch of latent vectors that we will use to visualize
+    #  the progression of the generator
+    fixed_noise = torch.randn(64, nz, 1, 1, device=device)
+
+    # Establish convention for real and fake labels during training
+    real_label = 1.0
+    fake_label = 0.0
+
+    # Set up Adam optimizers for both G and D
+    optimizer_d = optim.Adam(discriminator.parameters(), lr=lr, betas=(beta1, 0.999))
+    optimizer_g = optim.Adam(generator.parameters(), lr=lr, betas=(beta1, 0.999))
+
+    # Lists to keep track of progress
+    losses_g = []
+    losses_d = []
+    iteration = 0
+
+    # Training loop
+    for epoch in range(num_epochs):
+        for i, data in enumerate(dataloader, 0):
```

Comment:
tbh, it is confusing, looking up and down and the two training loops with and without fabric look very the same so as a user it not clear what is the fabric benefit... :/

Reply:
Yes, Fabric does not significantly change the code. There is not much boilerplate code in this example, so Fabric does very little for the user here. It is however possible to scale the code more easily, in ways that would otherwise require more boilerplate code to be added to the existing raw pytorch example.